byzcoin/contracts/deferred.go

```diff
@@ -0,0 +1,379 @@
+
+		instructionIDs := make([][]byte, len(c.DeferredData.ProposedTransaction.Instructions))
+
+		for i, proposedInstr := range c.DeferredData.ProposedTransaction.Instructions {
+
+			// In case it goes well, we want to return the proposed Tx InstanceID
+			instructionIDs[i] = proposedInstr.DeriveID("").Slice()
+
+			instructionType := proposedInstr.GetType()
+
+			var contractID string
+			switch instructionType {
+			case byzcoin.SpawnType:
+				contractID = proposedInstr.Spawn.ContractID
+			case byzcoin.InvokeType:
+				contractID = proposedInstr.Invoke.ContractID
+			case byzcoin.DeleteType:
+				contractID = proposedInstr.Delete.ContractID
+			}
+
+			fn, exists := c.s.GetContractConstructor(contractID)
+			if !exists {
+				return nil, nil, errors.New("Couldn't get the root function")
+			}
+			rootInstructionBuff, err := protobuf.Encode(&proposedInstr)
+			if err != nil {
+				return nil, nil, errors.New("Couldn't encode the root instruction buffer")
+			}
+			contract, err := fn(rootInstructionBuff)
+			if err != nil {
+				return nil, nil, errors.New("Couldn't get the root contract")
+			}
+			err = contract.VerifyDeferedInstruction(rst, proposedInstr, c.DeferredData.InstructionHashes[i])
+			if err != nil {
+				return nil, nil, fmt.Errorf("Verifying the instruction failed: %s", err)
+			}
+
+			var stateChanges []byzcoin.StateChange
+			switch instructionType {
+			case byzcoin.SpawnType:
+				stateChanges, _, err = contract.Spawn(rst, proposedInstr, coins)
+			case byzcoin.InvokeType:
+				stateChanges, _, err = contract.Invoke(rst, proposedInstr, coins)
+			case byzcoin.DeleteType:
+				stateChanges, _, err = contract.Delete(rst, proposedInstr, coins)
+
+			}
+
+			if err != nil {
+				return nil, nil, fmt.Errorf("Error while executing an instruction: %s", err)
+			}
+			sc = append(sc, stateChanges...)
+
+		}
+
+		c.DeferredData.ExecResult = instructionIDs
+		resultBuf, err2 := protobuf.Encode(&c.DeferredData)
+		if err2 != nil {
+			return nil, nil, errors.New("Couldn't encode the result")
+		}
+		sc = append(sc, byzcoin.NewStateChange(byzcoin.Update, inst.InstanceID,
+			ContractDeferredID, resultBuf, darcID))
+		return
+	default:
+		return nil, nil, errors.New("Deferred contract can only addProof and execProposedTx")
+	}
+}
+
+func (c *contractDeferred) Delete(rst byzcoin.ReadOnlyStateTrie, inst byzcoin.Instruction, coins []byzcoin.Coin) (sc []byzcoin.StateChange, cout []byzcoin.Coin, err error) {
+	cout = coins
+
+	// Find the darcID for this instance.
+	var darcID darc.ID
+	_, _, _, darcID, err = rst.GetValues(inst.InstanceID.Slice())
+	if err != nil {
+		return
+	}
+
+	sc = byzcoin.StateChanges{
+		byzcoin.NewStateChange(byzcoin.Remove, inst.InstanceID, ContractDeferredID, nil, darcID),
+	}
+	return
+}
+
+// VerifyInstruction overrides the basic VerifyInstruction
+func (c *contractDeferred) VerifyInstruction(rst byzcoin.ReadOnlyStateTrie, instr byzcoin.Instruction, ctxHash []byte) error {
```

Comment:
i'm not 100% sure whether what's in this function is what we think `VerifyInstruction` should be doing. if you look at the interface documentation, `VerifyInstruction` should only care whether the right signatures are there for the darc, it shouldn't have any other functionality. in my opinion these contract specific verification should go into spawn/invoke. what do you think @jeffallen

Reply:
Makes sense. Moved it to the spawn instruction.